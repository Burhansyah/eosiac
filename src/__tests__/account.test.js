const Account = require(`../account`)
const existingAccount = require(`./account.json`)

describe(`account.create`, () => {
    const env = {
        accounts_manager: `accntmngr`,
    }

    it(`returns new account action on empty current state`, async () => {
        expect.assertions(1)
        const accountConfig = {
            auth: {
                owner: {
                    permissions: [`EOS6z98haKv12d8vAJbDUkXmpK3PaKvf3Dv1NApo9MCxJ8oNWssDi`],
                },
                active: {
                    threshold: 2,
                    permissions: [
                        `EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x`,
                        `dapptoken@eosio.code 1`,
                        `wait@600 1`,
                    ],
                },
            },
        }
        const a = new Account(`test`, accountConfig)
        a.currentState = {}
        const actions = await a.create({env})
        expect(actions).toMatchSnapshot()
    })

    it(`returns no action on existing currentState`, async () => {
        expect.assertions(1)
        const desiredState = {
            auth: {
                owner: {
                    permissions: [`EOS6z98haKv12d8vAJbDUkXmpK3PaKvf3Dv1NApo9MCxJ8oNWssDi`],
                },
                active: {
                    threshold: 2,
                    permissions: [
                        `EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x`,
                        `dapptoken@eosio.code 1`,
                        `wait@600 1`,
                    ],
                },
            },
        }
        const a = new Account(`test`, desiredState)
        a.currentState = existingAccount
        const actions = await a.create({env})
        expect(actions).toBeFalsy()
    })
})

describe(`account.updateAuth`, () => {
    const env = {
        accounts_manager: `accntmngr`,
    }

    it(`throws when account does not exist yet`, async () => {
        expect.assertions(1)
        const accountConfig = {
            auth: {
                owner: {
                    permissions: [`EOS6z98haKv12d8vAJbDUkXmpK3PaKvf3Dv1NApo9MCxJ8oNWssDi`],
                },
                active: {
                    threshold: 2,
                    permissions: [
                        `EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x`,
                        `dapptoken@eosio.code 1`,
                        `wait@600 1`,
                    ],
                },
            },
        }
        const a = new Account(`test`, accountConfig)
        a.currentState = {}
        expect(a.updateAuth({env})).rejects.toThrow(/does not exist/i)
    })

    it(`returns no actions to run when used on same unordered permissions`, async () => {
        expect.assertions(1)
        const accountConfig = {
            auth: {
                owner: {
                    permissions: [`EOS6z98haKv12d8vAJbDUkXmpK3PaKvf3Dv1NApo9MCxJ8oNWssDi`],
                },
                active: {
                    threshold: 2,
                    permissions: [
                        `EOS84UkWvuV4tM8s7ose5wwF8MReUVBCwtSJHUFMdUxBu2B2cTm32`,
                        `EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x`,
                        `dapptoken@eosio.code 1`,
                        `wait@600 1`,
                    ],
                },
            },
        }
        const a = new Account(`test`, accountConfig)
        a.currentState = existingAccount
        const actions = await a.updateAuth({env})
        expect(actions).toMatchSnapshot()
    })

    it(`returns updateauth action when used with new permission`, async () => {
        expect.assertions(1)
        const accountConfig = {
            auth: {
                owner: {
                    permissions: [
                        `EOS6z98haKv12d8vAJbDUkXmpK3PaKvf3Dv1NApo9MCxJ8oNWssDi`,
                        `EOS7uSfs552n9hN1U2vidQybeP6dqx2V4qLEB5BdtEGPM3kyhGDnd`,
                    ],
                },
                active: {
                    threshold: 3,
                    permissions: [
                        `EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x`,
                        `EOS84UkWvuV4tM8s7ose5wwF8MReUVBCwtSJHUFMdUxBu2B2cTm32`,
                        `dapptoken@eosio.code 1`,
                        `wait@600 1`,
                    ],
                },
                newnewnew: {
                    threshold: 2,
                    permissions: [`test@active`, `wait@1200 1`],
                },
            },
        }
        const a = new Account(`test`, accountConfig)
        a.currentState = existingAccount
        const actions = await a.updateAuth({env})
        expect(actions).toMatchSnapshot()
    })
})
