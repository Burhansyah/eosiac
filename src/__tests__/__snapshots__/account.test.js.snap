// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account.create returns new account action on empty current state 1`] = `
Object {
  "account": "eosio",
  "authorization": Array [
    Object {
      "actor": "accntmngr",
      "permission": "active",
    },
  ],
  "data": Object {
    "active": Object {
      "accounts": Array [
        Object {
          "permission": Object {
            "actor": "dapptoken",
            "permission": "eosio.code",
          },
          "weight": 1,
        },
      ],
      "keys": Array [
        Object {
          "key": "EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x",
          "weight": 1,
        },
      ],
      "parent": "owner",
      "threshold": 2,
      "waits": Array [
        Object {
          "wait_sec": 600,
          "weight": 1,
        },
      ],
    },
    "creator": "accntmngr",
    "name": "test",
    "owner": Object {
      "accounts": Array [],
      "keys": Array [
        Object {
          "key": "EOS6z98haKv12d8vAJbDUkXmpK3PaKvf3Dv1NApo9MCxJ8oNWssDi",
          "weight": 1,
        },
      ],
      "parent": "",
      "threshold": 1,
      "waits": Array [],
    },
  },
  "name": "newaccount",
}
`;

exports[`account.updateAuth returns updateauth action when used with new permission 1`] = `
Array [
  Object {
    "account": "eosio",
    "authorization": Array [
      Object {
        "actor": "test",
        "permission": "owner",
      },
    ],
    "data": Object {
      "account": "test",
      "auth": Object {
        "accounts": Array [],
        "keys": Array [
          Object {
            "key": "EOS6z98haKv12d8vAJbDUkXmpK3PaKvf3Dv1NApo9MCxJ8oNWssDi",
            "weight": 1,
          },
          Object {
            "key": "EOS7uSfs552n9hN1U2vidQybeP6dqx2V4qLEB5BdtEGPM3kyhGDnd",
            "weight": 1,
          },
        ],
        "parent": "",
        "threshold": 1,
        "waits": Array [],
      },
      "parent": "",
      "permission": "owner",
    },
    "name": "updateauth",
  },
  Object {
    "account": "eosio",
    "authorization": Array [
      Object {
        "actor": "test",
        "permission": "owner",
      },
    ],
    "data": Object {
      "account": "test",
      "auth": Object {
        "accounts": Array [
          Object {
            "permission": Object {
              "actor": "dapptoken",
              "permission": "eosio.code",
            },
            "weight": 1,
          },
        ],
        "keys": Array [
          Object {
            "key": "EOS5nYs7LDcFcVtFmnYrhYQv9Z8DPDd1r4MJ8na2MvcPPJkXwzM3x",
            "weight": 1,
          },
          Object {
            "key": "EOS84UkWvuV4tM8s7ose5wwF8MReUVBCwtSJHUFMdUxBu2B2cTm32",
            "weight": 1,
          },
        ],
        "parent": "owner",
        "threshold": 3,
        "waits": Array [
          Object {
            "wait_sec": 600,
            "weight": 1,
          },
        ],
      },
      "parent": "owner",
      "permission": "active",
    },
    "name": "updateauth",
  },
  Object {
    "account": "eosio",
    "authorization": Array [
      Object {
        "actor": "test",
        "permission": "owner",
      },
    ],
    "data": Object {
      "account": "test",
      "auth": Object {
        "accounts": Array [
          Object {
            "permission": Object {
              "actor": "test",
              "permission": "active",
            },
            "weight": 1,
          },
        ],
        "keys": Array [],
        "parent": "owner",
        "threshold": 2,
        "waits": Array [
          Object {
            "wait_sec": 1200,
            "weight": 1,
          },
        ],
      },
      "parent": "owner",
      "permission": "newnewnew",
    },
    "name": "updateauth",
  },
]
`;

exports[`account.updateBandwidth returns both delegatebw & undelegatebw action when having more NET but less CPU 1`] = `
Array [
  Object {
    "account": "eosio.token",
    "authorization": Array [
      Object {
        "actor": "fundsmngr",
        "permission": "active",
      },
    ],
    "data": Object {
      "asset": "90.0000 EOS",
      "from": "fundsmngr",
      "memo": "Fund CPU/NET staking",
      "to": "test",
    },
    "name": "transfer",
  },
  Object {
    "account": "eosio",
    "authorization": Array [
      Object {
        "actor": "test",
        "permission": "active",
      },
    ],
    "data": Object {
      "from": "test",
      "receiver": "test",
      "unstake_cpu_quantity": "0.0000 EOS",
      "unstake_net_quantity": "80.0000 EOS",
    },
    "name": "undelegatebw",
  },
  Object {
    "account": "eosio",
    "authorization": Array [
      Object {
        "actor": "test",
        "permission": "active",
      },
    ],
    "data": Object {
      "from": "test",
      "receiver": "test",
      "stake_cpu_quantity": "90.0000 EOS",
      "stake_net_quantity": "0.0000 EOS",
      "transfer": false,
    },
    "name": "delegatebw",
  },
]
`;

exports[`account.updateBandwidth returns single delegatebw action when having less NET/CPU 1`] = `
Array [
  Object {
    "account": "eosio.token",
    "authorization": Array [
      Object {
        "actor": "fundsmngr",
        "permission": "active",
      },
    ],
    "data": Object {
      "asset": "120.0000 EOS",
      "from": "fundsmngr",
      "memo": "Fund CPU/NET staking",
      "to": "test",
    },
    "name": "transfer",
  },
  Object {
    "account": "eosio",
    "authorization": Array [
      Object {
        "actor": "test",
        "permission": "active",
      },
    ],
    "data": Object {
      "from": "test",
      "receiver": "test",
      "stake_cpu_quantity": "50.0000 EOS",
      "stake_net_quantity": "70.0000 EOS",
      "transfer": false,
    },
    "name": "delegatebw",
  },
]
`;

exports[`account.updateBandwidth returns single undelegatebw action when having more NET/CPU 1`] = `
Array [
  Object {
    "account": "eosio",
    "authorization": Array [
      Object {
        "actor": "test",
        "permission": "active",
      },
    ],
    "data": Object {
      "from": "test",
      "receiver": "test",
      "unstake_cpu_quantity": "100.0000 EOS",
      "unstake_net_quantity": "80.0000 EOS",
    },
    "name": "undelegatebw",
  },
]
`;

exports[`account.updateRam returns updateRam action when not enough RAM 1`] = `
Object {
  "account": "eosio",
  "authorization": Array [
    Object {
      "actor": "rammngr",
      "permission": "active",
    },
  ],
  "data": Object {
    "bytes": 154793,
    "payer": "rammngr",
    "receiver": "test",
  },
  "name": "buyrambytes",
}
`;
